src/Router.js

```diff
@@ -352,26 +362,28 @@ class ComponentScreen extends RequestScreen {
 	}
 
 	/**
-	 * Calls the handler specified by the router's `beforeDeactivateHandler`
-	 * state property.
-	 * @return {?boolean}
+	 * @inheritDoc
+	 */
+	beforeActivate() {
+		let handler = this.router.beforeActivateHandler;
+		if (handler) {
+			if (core.isString(handler)) {
+				handler = this.resolveHandler_(handler, this.router.component);
+			}
+			return handler();
+		}
+	}
+
+	/**
+	 * @inheritDoc
 	 */
 	beforeDeactivate() {
-		const handler = this.router.beforeDeactivateHandler;
+		let handler = this.router.beforeDeactivateHandler;
 		if (handler) {
 			if (core.isString(handler)) {
-				const comp = this.router.getRouteComponent();
-				if (comp && core.isFunction(comp[handler])) {
-					return comp[handler]();
-				} else {
-					const compName = getFunctionName(comp);
-					throw new Error(
-						`No function named "${handler}" exists inside ${compName}.`
-					);
-				}
-			} else {
-				return handler();
+				handler = this.resolveHandler_(handler);
```

Comment:
Shouldn't this take `this.router.component`?

Reply:
If you don't pass the second arg of `resolveHandler_`, it will resolve to `this.router.getRouteComponent()` which is what we want for `beforeDeactivateHandler`.

The reason for the difference is that `beforeActivateHandler` is invoked before the component is actually rendered, which is why we're just looking for a static method on the class (`this.router.component`) rather then the already rendered component.

There is some more explanation in the original PR I closed https://github.com/metal/metal-router/pull/29.

Reply:
You're right. I actually read that and immediately forgot 😂 

What if we...
- Remove the implicit logic from `resolveHandler_`, so `component` is a required attribute
- Change the calling code here to pass in `this.router.getRouteComponent()`
- Add a line comment on both calling methods explaining why that specific component and no other is being passed down so we won't forget in the future

Reply:
Sounds good!